app/history.py

```diff
@@ -19,44 +19,81 @@
 import app.log
 import cPickle as pickle
 import os
+import hashlib
+import time
+import app.prefs
 
-data = {}
-path = None
-
-def get(keyPath, default=None):
-  global data
-  cursor = data
-  for i in keyPath[:-1]:
-    cursor = cursor.setdefault(i, {})
-  return cursor.get(keyPath[-1], default)
-
-def set(keyPath, value):
-  global data
-  cursor = data
-  for i in keyPath[:-1]:
-    cursor = cursor.setdefault(i, {})
-  cursor[keyPath[-1]] = value
-  #assert get(keyPath) == value
-
-def loadUserHistory(filePath):
-  global data, path
-  try:
-    path = filePath
-    if os.path.isfile(path):
-      with open(path, "rb") as file:
-        data = pickle.load(file)
-    else:
-      data = {}
-  except Exception, e:
-    app.log.error('exception')
-    data = {}
+userHistory = {}
+pathToHistory = app.prefs.prefs['userData'].get('historyPath')
+
+def loadUserHistory(filePath, historyPath=pathToHistory):
+  global userHistory, pathToHistory
+  pathToHistory = historyPath
+  if os.path.isfile(historyPath):
+    with open(historyPath, 'rb') as file:
+      userHistory = pickle.load(file)
 
-def saveUserHistory():
-  global data, path
+def saveUserHistory(fileInfo, fileHistory, historyPath=pathToHistory):
+  """
+  Args:
+    fileInfo (tuple): Contains (filePath, lastChecksum, lastFileSize).
+    fileHistory (dict): The history of the file that the user wants to save.
+    historyPath (str): The path to the user's saved history.
+
+  Returns:
+    None
+  """
+  global userHistory, pathToHistory
+  filePath, lastChecksum, lastFileSize = fileInfo
   try:
-    if path is not None:
-      with open(path, "wb") as file:
-        pickle.dump(data, file)
+    if historyPath is not None:
+      pathToHistory = historyPath
+      userHistory.pop((lastChecksum, lastFileSize), None)
+      newChecksum, newFileSize = getFileInfo(filePath)
+      userHistory[(newChecksum, newFileSize)] = fileHistory
+      with open(historyPath, 'wb') as file:
+        pickle.dump(userHistory, file)
       app.log.info('wrote pickle')
   except Exception, e:
     app.log.error('exception')
+
+def getFileHistory(filePath):
+  checksum, fileSize = getFileInfo(filePath)
+  fileHistory = userHistory.get((checksum, fileSize), {})
+  fileHistory['adate'] = time.time()
+  return fileHistory
+
+def getFileInfo(filePath):
+  """
+  Args:
+    filePath (str): The absolute path to the file.
+
+  Returns:
+    A tuple containing the checksum and size of the file.
+  """
+  try:
+    checksum = calculateChecksum(filePath)
+    fileSize = os.stat(filePath).st_size
+    return (checksum, fileSize)
+  except:
+    return (None, 0)
+
+def calculateChecksum(filePath):
+  app.log.info("Calculate checksum of the current file")
+  hasher = hashlib.sha512()
+  with open(filePath, 'rb') as file:
+    hasher.update(file.read())
```

Comment:
Would it be reasonable to use the file data read here:
https://github.com/google/ci_edit/blob/master/app/actions.py#L542
to avoid reading the file twice?